### PR TITLE
PhiMoE

### PR DESCRIPTION
**Pull request by garg-amit**
# What does this PR do?

Integrates `PhiMoE` into transformers. https://huggingface.co/microsoft/Phi-3.5-MoE-instruct

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @gante

### CODE REVIEW

**Comment by garg-amit** [comment]
@ArthurZucker @gante can I please get a review?

**Comment by merryHunter** [comment]
Hi, it seems to be a very important and awaited PR!:) Other frameworks are willing to integrate MoE too, like in litgpt [https://github.com/Lightning-AI/litgpt/issues/1686](https://github.com/Lightning-AI/litgpt/issues/1686).

File: docs/source/en/perf_infer_gpu_one.md [hunk 1]
```diff
@@ -77,6 +77,8 @@ FlashAttention-2 is currently supported for the following architectures:
 * [OPT](https://huggingface.co/docs/transformers/model_doc/opt#transformers.OPTModel)
 * [Phi](https://huggingface.co/docs/transformers/model_doc/phi#transformers.PhiModel)
 * [Phi3](https://huggingface.co/docs/transformers/model_doc/phi3#transformers.Phi3Model)
+* [PhiMoE](https://huggingface.co/docs/transformers/model_doc/phimoe#transformers.PhiMoEModel)
+* [SigLIP](https://huggingface.co/docs/transformers/model_doc/siglip)
```
Comment (ArthurZucker) [inline docs/source/en/perf_infer_gpu_one.md:81]: unrelated change it seems no?

File: src/transformers/__init__.py [hunk 2]
```diff
@@ -5441,6 +5450,7 @@
     )
     from .models.phi import PhiConfig
     from .models.phi3 import Phi3Config
+    from .models.phimoe import PhiMoEConfig
```
Comment (ArthurZucker) [inline src/transformers/__init__.py:5453]:
```suggestion
    from .models.phimoe import PhimoeConfig
```
let's follow `OlmoeForCausalLM` convention of camel cased

File: src/transformers/models/phimoe/configuration_phimoe.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+from ...configuration_utils import PretrainedConfig
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+PHIMOE_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "microsoft/Phi-3.5-MoE-instruct": "https://huggingface.co/microsoft/Phi-3.5-MoE-instruct/resolve/main/config.json",
+}
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/configuration_phimoe.py:27]:
```suggestion
```

File: src/transformers/models/phimoe/configuration_phimoe.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+            Whether or not the router logits should be returned by the model. Enabeling this will also
+            allow the model to output the auxiliary loss. See [here]() for more details
+        router_aux_loss_coef (`float`, *optional*, defaults to 0.001):
+            The aux loss factor for the total loss.
+        router_jitter_noise (`float`, *optional*, defaults to 0.01):
+            Amount of noise to add to the router.
+        input_jitter_noise (`float`, *optional*, defaults to 0.0): Input jitter noise
+        attention_bias (`bool`, *optional*, defaults to `False`): Attention bias
+        lm_head_bias (`bool`, *optional*, defaults to `False`): LM head bias
+    ```python
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/configuration_phimoe.py:104]:
```suggestion

    Example:

    ```python
```

File: src/transformers/models/phimoe/configuration_phimoe.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                f"`rope_scaling`'s long_factor field must have length {self.hidden_size // self.num_attention_heads // 2}, got {len(rope_scaling_long_factor)}"
+            )
+        if not isinstance(rope_scaling_short_mscale, (int, float)):
+            raise ValueError(f"`rope_scaling`'s short_mscale field must be a number, got {rope_scaling_short_mscale}")
+        if not isinstance(rope_scaling_long_mscale, (int, float)):
+            raise ValueError(f"`rope_scaling`'s long_mscale field must be a number, got {rope_scaling_long_mscale}")
+        if not isinstance(original_max_position_embeddings, int):
+            raise ValueError(
+                f"`rope_scaling`'s original_max_position_embeddings field must be an integer, got {original_max_position_embeddings}"
+            )
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/configuration_phimoe.py:236]: Let's use this one: https://github.com/huggingface/transformers/blob/7e638ef2b8650aaa3e3a8e575bb63af262a43d95/src/transformers/models/olmoe/configuration_olmoe.py#L167-L171
```suggestion
```

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 6]
```diff
@@ -0,0 +1,1850 @@
...
+)
+from ...utils.import_utils import is_torch_fx_available
+from .configuration_phimoe import PhiMoEConfig
+
+
+if is_flash_attn_2_available():
+    from flash_attn import flash_attn_func, flash_attn_varlen_func
+    from flash_attn.bert_padding import index_first_axis, pad_input, unpad_input  # noqa
+
+    _flash_supports_window_size = "window_size" in list(inspect.signature(flash_attn_func).parameters)
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:58]: This was updated to https://github.com/huggingface/transformers/blob/7e638ef2b8650aaa3e3a8e575bb63af262a43d95/src/transformers/modeling_flash_attention_utils.py#L183

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 7]
```diff
@@ -0,0 +1,1850 @@
...
+            mask_length = attention_mask.shape[-1]
+            padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :]
+            padding_mask = padding_mask == 0
+            causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
+                padding_mask, min_dtype
+            )
+
+    return causal_mask
+
+
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:127]:
```suggestion
# Copied from transformers.models.jetmoe.modeling_jetmoe.load_balancing_loss_func
```

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 8]
```diff
@@ -0,0 +1,1850 @@
...
+        inv_freq = 1.0 / (
+            rescale_factors * (self.base ** (torch.arange(0, self.dim, 2).float().to(x.device) / self.dim))
+        )
+
+        t = torch.arange(seq_len, device=x.device, dtype=torch.float32)
+        freqs = torch.outer(t, inv_freq)
+
+        emb = torch.cat((freqs, freqs), dim=-1)
+        return (emb.cos() * mscale).to(x.dtype), (emb.sin() * mscale).to(x.dtype)
+
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:284]:
```suggestion
```
new rope scaling should be added to this: https://github.com/huggingface/transformers/blob/7e638ef2b8650aaa3e3a8e575bb63af262a43d95/src/transformers/modeling_rope_utils.py#L353-L360

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 9]
```diff
@@ -0,0 +1,1850 @@
...
+            self.rotary_emb = PhiMoERotaryEmbedding(
+                self.head_dim,
+                max_position_embeddings=self.max_position_embeddings,
+                base=self.rope_theta,
+            )
+        else:
+            scaling_type = self.config.rope_scaling["type"]
+            if scaling_type == "longrope":
+                self.rotary_emb = Phi3LongRoPEScaledRotaryEmbedding(self.head_dim, self.config)
+            else:
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:385]: to remove

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 10]
```diff
@@ -0,0 +1,1850 @@
...
+    num_key_value_heads, seqlen, head_dim) to (batch, num_attention_heads, seqlen, head_dim)
+    """
+    batch, num_key_value_heads, slen, head_dim = hidden_states.shape
+    if n_rep == 1:
+        return hidden_states
+    hidden_states = hidden_states[:, :, None, :, :].expand(batch, num_key_value_heads, n_rep, slen, head_dim)
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
+
+
+class PhiMoEAttention(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:334]: should be completely copied form gemm a

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 11]
```diff
@@ -0,0 +1,1850 @@
...
+
+        attn_output = self.o_proj(attn_output)
+
+        if not output_attentions:
+            attn_weights = None
+
+        return attn_output, attn_weights, past_key_value
+
+
+class PhiMoEFlashAttention2(PhiMoEAttention):
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:475]: same here

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 12]
```diff
@@ -0,0 +1,1850 @@
...
+            query_layer, indices_q, cu_seqlens_q, max_seqlen_in_batch_q = unpad_input(query_layer, attention_mask)
+
+        return (
+            query_layer,
+            key_layer,
+            value_layer,
+            indices_q,
+            (cu_seqlens_q, cu_seqlens_k),
+            (max_seqlen_in_batch_q, max_seqlen_in_batch_k),
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:765]:
```suggestion
```

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 13]
```diff
@@ -0,0 +1,1850 @@
...
+
+
+PHIMOE_ATTENTION_CLASSES = {
+    "eager": PhiMoEAttention,
+    "flash_attention_2": PhiMoEFlashAttention2,
+    "sdpa": PhiMoESdpaAttention,
+}
+
+
+class PhiMoEBlockSparseTop2MLP(nn.Module):
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:862]: Euuu this looks like a simple up gate down mlp, copied form gemma as well no?

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 14]
```diff
@@ -0,0 +1,1850 @@
...
+        )
+
+        return (
+            grad_at_scores_expaned,
+            None,
+            None,
+            None,
+            None,
+        )
+
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:924]: not sure I understand why we have this. If required we need to document why, we need better naming as well

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 15]
```diff
@@ -0,0 +1,1850 @@
...
+    masked_gates = scores.masked_fill(mask_logits_threshold, float("-inf"))
+    if training:
+        selected_experts = (
+            (
+                masked_gates
+                - torch.empty_like(masked_gates, memory_format=torch.legacy_contiguous_format).exponential_().log()
+            )
+            .max(dim=-1)[1]
+            .unsqueeze(-1)
+        )  # gumbel sampling, more robust than than the multinomial method
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:947]: let's doccument these kind of stuff, why are we doing this, which paper / experiments, what is this gumbel sampling etc

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 16]
```diff
@@ -0,0 +1,1850 @@
...
+            grad_at_scores_expaned,
+            None,
+            None,
+            None,
+            None,
+        )
+
+
+def sparsemixer(scores, top_k, jitter_eps, training):
+    assert top_k == 2
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:927]: this is very weird, we should either use the mixtral / switch transformers formulation, if they are equivalent, or we build form the latter, but we need to make this clean as it is a core component of the model.

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 17]
```diff
@@ -0,0 +1,1850 @@
...
+        for expert_idx in range(self.num_experts):
+            expert_layer = self.experts[expert_idx]
+            idx, top_x = torch.where(expert_mask[expert_idx])
+
+            if top_x.shape[0] == 0:
+                continue
+
+            # in torch it is faster to index using lists than torch tensors
+            top_x_list = top_x.tolist()
+            idx_list = idx.tolist()
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:1099]: I think this was removed form the mixtral modeling so to update as well

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 18]
```diff
@@ -0,0 +1,1850 @@
...
+        # gating
+        self.gate = nn.Linear(self.hidden_dim, self.num_experts, bias=False)
+
+        self.experts = nn.ModuleList([PhiMoEBlockSparseTop2MLP(config) for _ in range(self.num_experts)])
+
+        # Jitter parameters
+        self.router_jitter_noise = config.router_jitter_noise
+        self.input_jitter_noise = config.input_jitter_noise
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:1064]: most probably copied from mixtral

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 19]
```diff
@@ -0,0 +1,1850 @@
...
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.embed_tokens = value
+
+    # Ignore copy
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:1325]: we should not ignore copy here!

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 20]
```diff
@@ -0,0 +1,1850 @@
...
+            loss=loss,
+            aux_loss=aux_loss,
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+            router_logits=outputs.router_logits,
+        )
+
+    def prepare_inputs_for_generation(
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:1642]: copied from phi3 no?

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 21]
```diff
@@ -0,0 +1,1850 @@
...
+        return MoeModelOutputWithPast(
+            last_hidden_state=hidden_states,
+            past_key_values=next_cache,
+            hidden_states=all_hidden_states,
+            attentions=all_self_attns,
+            router_logits=all_router_logits,
+        )
+
+
+class PhiMoEForCausalLM(PhiMoEPreTrainedModel):
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:1490]: this can also be copied from phi3 no?

**Review by ArthurZucker — COMMENTED** [review]
Thanks! 
Let's go with camel cased classes, If we want to be compile compatible we need to have a script conversion and use the formulation from gpt fast moe with a version implemented here: https://github.com/huggingface/transformers/pull/30793/files#diff-733ab0a772c69f78b1d8ed361e6ae1fda7243652887aed0bab5d3ecf07794c01R789

Lot's of stuff seems similar to phi3 so we can probably copy from it!

**Comment by ArthurZucker** [comment]
TLDR, overall the mixer needs to be properly documented and written to be more understandable!

File: docs/source/en/perf_infer_gpu_one.md (hunk 1 above)
Comment (garg-amit) [reply docs/source/en/perf_infer_gpu_one.md]: Removed SigLIP

File: src/transformers/__init__.py (hunk 2 above)
Comment (garg-amit) [reply src/transformers/__init__.py]: Fixed

File: src/transformers/models/phimoe/configuration_phimoe.py (hunk 3 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/configuration_phimoe.py]: Removed

File: src/transformers/models/phimoe/configuration_phimoe.py (hunk 5 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/configuration_phimoe.py]: Updated it to use the `rope_config_validation` fn instead

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 6 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Updated

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 7 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py:121]: Added the comment

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 11 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: It's copied from `MixtralFlashAttention2`. Added the comment

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 12 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Removed

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 13 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: yeah, it's copied from `transformers.models.mixtral.modeling_mixtral.MixtralBlockSparseTop2MLP`. Fixed

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 14 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: It’s needed in the `sparsemixer`. I've renamed the class and added a docstring.

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 17 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Updated

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 18 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py:921]: It's a bit different as it uses the `sparsemixer`.

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 20 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py:1515]: yeah, added the following comment `# copied from transformers.models.phi3.modeling_phi3.prepare_inputs_for_generation`

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 15 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Sure, I've added the paper link

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 16 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py:789]: It is different from the above and is based on the following paper: https://arxiv.org/pdf/2409.12136. I’ve added this info in the docstring.

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 21 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: It is quite similar to the `MixtralForCausalLM` class, except for the warning in the `forward` function.

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 8 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Modified the longrope implementation to use short factors for shorter sequences when seq_len is less than original_max_position_embeddings.

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 9 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Removed the `Phi3LongRoPEScaledRotaryEmbedding` class

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 10 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: It is similar to the `MixtralAttention` class but has some minor differences, such as `attention_bias`.

**Comment by garg-amit** [comment]
@ArthurZucker Thanks for reviewing the PR. I’ve refactored the code according to your suggestions, and it’s ready for another look. Also, the failing test case appears to be unrelated to this PR. Please let me know if it needs to be addressed.

**Comment by ArthurZucker** [comment]
Reviewing! 🤗

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 22]
```diff
@@ -0,0 +1,1716 @@
...
+    Args:
+        scores (torch.Tensor): Input scores tensor.
+        top_k (int): Number of top experts to select.
+        jitter_eps (float): Jitter epsilon for numerical stability.
+        training (bool): Flag indicating if the model is in training mode.
+
+    Returns:
+        Tuple[torch.Tensor, torch.Tensor]: Multiplier and selected experts tensors.
+    """
+    assert top_k == 2
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:724]: also let's raise an error rather than an assert!

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 23]
```diff
@@ -0,0 +1,1716 @@
...
+            None,
+            None,
+        )
+
+
+def sparsemixer(scores, top_k, jitter_eps, training):
+    """
+    Sparse mixer function to select top-k experts and compute multipliers.
+    Based on the paper: https://arxiv.org/pdf/2409.12136
+
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:714]:
```suggestion
  We first replace the T opK(·) function as random sampling of discrete variables
  in model training. Then, following Liu et al. (2023a) and Liu et al. (2023b), we apply Heun’s
  third order method to approximate the expert routing gradient and construct a modified
  back-propagation to give a mathematically sound gradient estimation for expert routing.
```

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 24]
```diff
@@ -0,0 +1,1716 @@
...
+        if self.training and self.input_jitter_noise > 0:
+            hidden_states *= torch.empty_like(hidden_states).uniform_(
+                1.0 - self.input_jitter_noise, 1.0 + self.input_jitter_noise
+            )
+        hidden_states = hidden_states.view(-1, hidden_dim)
+        router_logits = self.gate(hidden_states)
+
+        routing_weights, selected_experts = sparsemixer(
+            router_logits,
+            top_k=2,
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:872]: if it's hardcoded we can also just not put it!

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 25]
```diff
@@ -0,0 +1,1716 @@
...
+
+        self.self_attn = PHIMOE_ATTENTION_CLASSES[config._attn_implementation](config, layer_idx)
+
+        self.block_sparse_moe = PhimoeSparseMoeBlock(config)
+        self.input_layernorm = nn.LayerNorm(config.hidden_size, eps=config.rms_norm_eps, elementwise_affine=True)
+        self.post_attention_layernorm = nn.LayerNorm(
+            config.hidden_size, eps=config.rms_norm_eps, elementwise_affine=True
+        )
+
+    # copied from transformers.models.mixtral.modeling_mixtral.MixtralDecoderLayer.forward
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:919]:
```suggestion
    # Copied from transformers.models.mixtral.modeling_mixtral.MixtralDecoderLayer.forward
```

**Review by ArthurZucker — COMMENTED** [review]
LGTM, the only thing needed to merge: 
1. The `Copied` from need a capital letter
2. The core part needs a tad bit more doc as I said, why do we need a specific gradient computation (had to go through the paper to see that indeed you need a special gradient approx)
3. That part of the code is IMO less readable than the rest, but fine for now! 
THanks and sorry for the late revies!

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 24 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: @ArthurZucker I’ve removed `top_k` from here and instead created it as a keyword argument.

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 22 above)
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Fixed

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 26]
```diff
@@ -0,0 +1,1716 @@
...
+        )
+        self.v_proj = nn.Linear(
+            self.hidden_size, self.num_key_value_heads * self.head_dim, bias=self.config.attention_bias
+        )
+        self.o_proj = nn.Linear(self.num_heads * self.head_dim, self.hidden_size, bias=self.config.attention_bias)
+
+        self.rotary_emb = PhimoeRotaryEmbedding(
+            config=self.config,
+        )
+
```
[the remark this answers is not in the document]
Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Sure, moved it to the `PhimoeModel` class

**Comment by garg-amit** [comment]
@ArthurZucker Thanks for reviewing! I've addressed the comments and moved `PhimoeRotaryEmbedding` out of the `PhimoeAttention` class.

File: src/transformers/models/phimoe/modeling_phimoe.py [hunk 27]
```diff
@@ -0,0 +1,1709 @@
...
+        causal_mask = self._update_causal_mask(
+            attention_mask, inputs_embeds, cache_position, past_key_values, output_attentions
+        )
+
+        hidden_states = inputs_embeds
+
+        kv_seq_len = hidden_states.shape[-2]
+        if past_key_values is not None:
+            kv_seq_len += past_key_values.get_usable_length(kv_seq_len)
+        position_embeddings = self.rotary_emb(hidden_states, seq_len=kv_seq_len)
```
Comment (ArthurZucker) [inline src/transformers/models/phimoe/modeling_phimoe.py:1136]: pretty sure you should be using cache positions here! cache_position[0]!

**Review by ArthurZucker — APPROVED** [review]
Great work! Thanks for integrating this new model 🔥

File: src/transformers/models/phimoe/modeling_phimoe.py (hunk 27 above)
Comment (ArthurZucker) [reply src/transformers/models/phimoe/modeling_phimoe.py:1136]: it's the last nit!

Comment (garg-amit) [reply src/transformers/models/phimoe/modeling_phimoe.py]: Thanks for the suggestion! I've updated it to `cache_position[-1]+1` as `cache_position[0]` would return 0 when the kv cache is empty.

Comment (ArthurZucker) [reply src/transformers/models/phimoe/modeling_phimoe.py]: indeed! 🤗

**Comment by ArthurZucker** [comment]
Thanks everyone and @garg-amit for bearing with me! Congrats on the model release again 🤗
